libbpf-tools/funclatency.c

```diff
@@ -0,0 +1,370 @@
+			argp_usage(state);
+		}
+		env->units = MSEC;
+		break;
+	case 'u':
+		if (env->units != NSEC) {
+			warn("only set one of -m or -u\n");
+			argp_usage(state);
+		}
+		env->units = USEC;
+		break;
+	case 'd':
+		errno = 0;
+		duration = strtol(arg, NULL, 10);
+		if (errno || duration <= 0) {
+			warn("Invalid duration: %s\n", arg);
+			argp_usage(state);
+		}
+		env->duration = duration;
+		break;
+	case 'i':
+		errno = 0;
+		interval = strtol(arg, NULL, 10);
+		if (errno || interval <= 0) {
+			warn("Invalid interval: %s\n", arg);
+			argp_usage(state);
+		}
+		env->interval = interval;
+		break;
+	case 'T':
+		env->timestamp = true;
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case ARGP_KEY_ARG:
+		if (env->funcname) {
+			warn("Too many function names: %s\n", arg);
+			argp_usage(state);
+		}
+		env->funcname = arg;
+		break;
+	case ARGP_KEY_END:
+		if (!env->funcname) {
+			warn("Need a function to trace\n");
+			argp_usage(state);
+		}
+		if (env->duration) {
+			if (env->interval > env->duration)
+				env->interval = env->duration;
+			env->iterations = env->duration / env->interval;
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static const char *unit_str(void)
+{
+	switch (env.units) {
+	case NSEC:
+		return "nsec";
+	case USEC:
+		return "usec";
+	case MSEC:
+		return "msec";
+	};
+
+	return "bad units";
+}
+
+static void print_hist(int fd)
+{
+	unsigned int nr_cpus = libbpf_num_possible_cpus();
+	unsigned int hist[MAX_SLOTS] = {0};
+	uint64_t *count;
+
+	count = calloc(nr_cpus, sizeof(*count));
+	if (!count)
+		handle_error("calloc");
+
+	for (int i = 0; i < MAX_SLOTS; i++) {
+		if (bpf_map_lookup_elem(fd, &i, count))
+			handle_error("lookup");
+		hist[i] = 0;
+		for (int c = 0; c < nr_cpus; c++)
+			hist[i] += count[c];
+	}
+	free(count);
+
+	print_log2_hist(hist, MAX_SLOTS, unit_str());
+}
+
+static int attach_kprobes(struct funclatency_bpf *obj)
+{
+	struct bpf_link *l;
+
+	l = bpf_program__attach_kprobe(obj->progs.dummy_kprobe, false,
+				       env.funcname);
+	if (IS_ERR(l)) {
+		warn("failed to attach kprobe: %ld\n", PTR_ERR(l));
+		return -1;
+	}
+
+	l = bpf_program__attach_kprobe(obj->progs.dummy_kretprobe, true,
+				       env.funcname);
+	if (IS_ERR(l)) {
+		warn("failed to attach kretprobe: %ld\n",
+			PTR_ERR(l));
+		return -1;
+	}
+
+	return 0;
+}
+
+static int attach_uprobes(struct funclatency_bpf *obj)
+{
+	char *binary, *function, *bin_path;
+	off_t func_off;
+	int ret = -1;
+	struct bpf_link *l;
+
+	/*
+	 * TODO: libbpf cannot attach uprobes to all processes on all cpus.  It
+	 * only handles a specific pid (> 0) or the caller's pid (== 0) on any
+	 * cpu.  If you pass -1 (any pid), it will only trace on cpu 0.
+	 */
+	if (!env.pid) {
+		warn("Need a PID for uprobes (for now)\n");
+		return -1;
+	}
+
+	binary = strdup(env.funcname);
+	if (!binary)
+		handle_error("strdup");
+	function = strchr(binary, ':');
```

Comment:
probably worthwhile to leave a comment that binary is guaranteed to contain ':' (handled by attach_probes below)

Reply:
added a check for !function and warn() + return.